cherab/tools/inversions/opencl/sart_opencl.py

```diff
@@ -0,0 +1,279 @@
+    :param pyopencl.Device device: OpenCL device which will be used for computations. Default value: `device=None` (autoselect).
+    :param int block_size: Number of GPU threads per block. Must be the power of 2.
+        For the best performance try from 256 to 1024 for Nvidia (use 1024 on high-end GPUs),
+        from 64 to 256 for AMD and from 16 to 64 for Intel GPUs. Default value: `block_size=256`.
+    :param bool copy_column_major: If True, the two copies of geometry matrix will be stored in the GPU memory.
+        One in row-major order and the other one in column-major order. This provides much better performance of the
+        inversions but requires twice as much GPU memory. Default value: `copy_column_major=True`.
+    :param int block_size_row_maj: If `copy_column_major` is set to False, this parameter defines the number of GPU threads per block
+        in mat_vec_mult_row_maj() kernel used to calculate y_hat. Must be lower than `block_size`.
+        Default value: `block_size_row_maj=64` (optimal value for Nvidia GPUs).
+    :param bool use_atomic: If True, increases the number of thread blocks that can run in parallel with the help of atomic
+        operations (custom atomic add on floats). Set this to False, if the atomic operations are running slow
+        on your device (Nvidia GPUs before Kepler, some AMD APUs, some Intel GPUs). Default value: `use_atomic=True`.
+    :param int steps_per_thread: If `use_atomic` is set to True, this parameters defines the maximum number of loop steps performed
+        by the parallel threads in a single thread block. Default value: `steps_per_thread=64` (optimal for Nvidia GPUs).
+    """
+
+    def __init__(self, geometry_matrix, laplacian_matrix=None, device=None, block_size=256, copy_column_major=True, block_size_row_maj=64,
+                 use_atomic=True, steps_per_thread=64):
+        if not _has_pyopencl:
+            raise RuntimeError("The pyopencl module is required to use the SartOpencl() inversion class.")
+        if geometry_matrix.dtype != np.float32:  # converting geometry_matrix to float32 if needed
+            geometry_matrix = geometry_matrix.astype(np.float32)
+        self.m_detectors, self.n_sources = geometry_matrix.shape
+        cell_ray_densities = geometry_matrix.sum(0)
+        ray_lengths = geometry_matrix.sum(1)
+        device = device or device_select()
+        self.use_atomic = use_atomic
+        steps_per_thread = min(block_size, steps_per_thread)
+        steps_per_thread_row_maj = min(block_size_row_maj, steps_per_thread)
+
+        # creating OpenCL context
+        self.cl_context = cl.Context([device])
+
+        # reading and compiling OpenCL kernels
+        kernels_filename = 'sart_kernels_atomic.cl' if use_atomic else 'sart_kernels.cl'
+        kernel_source_file = os.path.join(os.path.dirname(os.path.realpath(__file__)), kernels_filename)
+        kernel_source = open(kernel_source_file).read()

```

Comment:
This gives me a ResourceWarning when running the OpenCL unit tests:
```
/home/jlovell/FUN/cherab/cherab_core/cherab/tools/inversions/opencl/sart_opencl.py:85: ResourceWarning: unclosed file <_io.TextIOWrapper name='/home/jlovell/FUN/cherab/cherab_core/cherab/tools/inversions/opencl/sart_kernels.cl' mode='r' encoding='UTF-8'>
  kernel_source = open(kernel_source_file).read()
ResourceWarning: Enable tracemalloc to get the object allocation traceback
```
Could you please properly close the file by wrapping in a with statement:
```python
with open(kernel_source_file, "r") as f:
    kernel_source = f.read()
```

Reply:
Thanks. I fixed this.